dusty/parallel.py

```diff
@@ -0,0 +1,47 @@
+"""Utilities for multithreaded parallel execution of tasks."""
+
+import threading
+import multiprocessing
+import multiprocessing.pool
+from contextlib import contextmanager
+from Queue import Queue
+
+from .log import log_to_client
+
+class TaskQueue(Queue, object):
+    """Executable task queue used for multithreaded execution of multiple
+    function calls. Concurrency is limited by `pool_size`."""
+    def __init__(self, pool_size):
+        super(TaskQueue, self).__init__()
+        self.pool = multiprocessing.pool.ThreadPool(pool_size)
+        self.errors = []
+
+    def enqueue_task(self, fn, *args, **kwargs):
+        self.put((fn, args, kwargs))
+
+    def _task_executor(self, fn, args, kwargs):
+        try:
+            fn(*args, **kwargs)
+        except Exception as e:
+            self.errors.append(e.message)
+
+    def execute(self):
+        while not self.empty():
+            fn, args, kwargs = self.get()
+            self.pool.apply_async(self._task_executor, args=(fn, args, kwargs))
+        self.pool.close()
```

Comment:
Since it looks like you can't reuse the pool after you've closed it, I think we should do one of the following:
1. Do something to indicate that TaskQueue should not invoked directly outside of parallel.py
2. Create the pool inside `execute`